cmd/curio/ffi.go

```diff
@@ -0,0 +1,116 @@
+package main
+
+import (
+	"context"
+	"encoding/gob"
+	"fmt"
+	"os"
+	"reflect"
+
+	"github.com/samber/lo"
+	"github.com/urfave/cli/v2"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/lotus/cmd/curio/deps"
+	"github.com/filecoin-project/lotus/lib/ffiselect"
+	"github.com/filecoin-project/lotus/storage/paths"
+	"github.com/filecoin-project/lotus/storage/sealer/ffiwrapper"
+	"github.com/filecoin-project/lotus/storage/sealer/storiface"
+)
+
+var ffiCmd = &cli.Command{
+	Name:   "ffi",
+	Hidden: true,
+	Flags: []cli.Flag{
+		layersFlag,
+	},
+	Action: func(cctx *cli.Context) (err error) {
+		output := os.NewFile(uintptr(3), "out")
+
+		defer func() {
+			if r := recover(); r != nil {
+				err = fmt.Errorf("panic: %v", r)
+			}
+			if err != nil {
+				gob.NewEncoder(output).Encode(ffiselect.ValErr{Val: nil, Err: err})
+			}
+		}()
+		var callInfo ffiselect.FFICall
+		if err := gob.NewDecoder(os.Stdin).Decode(&callInfo); err != nil {
+			return err
+		}
+
+		// wasteful, but *should* work
+		depnd, err := deps.GetDeps(cctx.Context, cctx)
+
+		w, err := ffiwrapper.New(&sectorProvider{index: depnd.Si, stor: depnd.LocalStore})
```

Comment:
This wrapper definitely shouldn't wrap around the ffiwrapper. It absolutely can't have it's own storage/path Index/Remote/Local instances, those must be in the same process as the scheduler, otwerwise we break storage reservations.

This command should allow for invoking methods one layer deeper, like [SealPreCommitPhase2](https://github.com/filecoin-project/lotus/blob/3d6ae227ef031c06f2c5062083b5d1059766db78/curiosrc/ffi/sdr_funcs.go#L259) - that will be much, much simpler as we won't need connect to the db (no GetDeps above) and interact with the storage index for anything. Also storage reservations won't need a rewrite.

The wrapper should be integrated in `curiosrc/ffi`, not in tasks; the curio ffi wrapper is meant to abstract interactions with the proofs library down to "run this algorithm on this sector-id now"

Reply:
I connected it at the FFI level, but I still like reflection so I made "ffidirect" under "ffiselect" as we will need to update both together. 

 I removed the dependency on LocalWorker and copied-over the related funcs to get access to where the edits were needed. That's why "package deps" was touched. 
This tech debt clean-up simplifies the mental model in this area and helps us avoid importing skeletons.